### PR TITLE
[TED] extractor rewrite

### PR DESCRIPTION
**Pull request by trassshhub**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

TED has restructured their site and a rewrite for TED extractor is required (#2343).

changes:
- rewrite talks and playlist handling
- added series handling
- removed handling /watch/ urls (was very old code and new watch pages no longer contains a single video as the old extractor expects)

I have used quite a bit of the old parsing code as some of them still matches the new json structure. There are parts of code which I am not sure if they are still necessary with their new player (e.g. audio_download and rtmp source), but since they are not fatal I chose to leave them as is.

### CODE REVIEW

File: yt_dlp/extractor/ted.py [hunk 1]
```diff
@@ -26,120 +24,63 @@ class TEDIE(InfoExtractor):
...
-            'title': 'The illusion of consciousness',
-            'description': ('Philosopher Dan Dennett makes a compelling '
-                            'argument that not only don\'t we understand our own '
-                            'consciousness, but that half the time our brains are '
-                            'actively fooling us.'),
-            'uploader': 'Dan Dennett',
-            'width': 853,
-            'duration': 1308,
+            'title': 'How to break down barriers and not accept limits',
+            'description': 'What can\'t Candace Parker do? A two-time NCAA champion, two-time Olympic gold medalist and two-time WNBA champion, Parker knows what it takes to fight for your dreams. In this inspiring talk, she shares what she\'s learned during a career spent not accepting limits -- and how her daughter taught her the best lesson of all. "Barrier breaking is about not staying in your lane and not being something that the world expects you to be," she says. "It\'s about not accepting limitations."',
```
Comment (pukkandan) [inline yt_dlp/extractor/ted.py:42]: too long. Use md5

File: yt_dlp/extractor/ted.py [hunk 2]
```diff
@@ -1,20 +1,18 @@
...
-import json
 import re
 
 from .common import InfoExtractor
 
 from ..compat import (
-    compat_str,
-    compat_urlparse
+    compat_str
 )
```
Comment (pukkandan) [inline yt_dlp/extractor/ted.py:8]: Since u r rewriting the whole thing anyway, change it to just `str`

File: yt_dlp/extractor/ted.py [hunk 3]
```diff
@@ -26,120 +24,63 @@ class TEDIE(InfoExtractor):
...
-            'tags': list,
-        },
-        'params': {
-            'skip_download': True,
+            'tags': ['personal growth', 'equality', 'activism', 'motivation', 'social change', 'sports'],
+            'uploader': 'Candace Parker',
+            'duration': 676.0,
+            'upload_date': '20220114',
+            'release_date': '20211201',
+            'thumbnail': 'https://pi.tedcdn.com/r/talkstar-photos.s3.amazonaws.com/uploads/fd2ff863-26bb-4a72-8fd4-8380a4a8c0b4/CandaceParker_2021W-embed.jpg',
```
Comment (pukkandan) [inline yt_dlp/extractor/ted.py:49]: is this URL stable? use a regex if not

File: yt_dlp/extractor/ted.py [hunk 4]
```diff
@@ -26,120 +24,63 @@ class TEDIE(InfoExtractor):
...
-            r'(?s)q\(\s*"\w+.init"\s*,\s*({.+?})\)\s*</script>',
-            webpage, 'info json')
-        return json.loads(info_json)
+            for entry in try_get(playlist, lambda x: x['videos']['nodes'], list):
+                if entry.get('__typename') == 'Video':
+                    playlist_entries.append(self.url_result(entry.get('canonicalUrl'), self.ie_key()))
+
+            return playlist_entries
+        else:
+            return []
```
Comment (pukkandan) [inline yt_dlp/extractor/ted.py:83]:
```suggestion
    def _parse_playlist(self, playlist):
        if not isinstance(playlist, dict):
            return []
        return [
            self.url_result(entry['canonicalUrl'], self.ie_key())
            for entry in try_get(playlist, lambda x: x['videos']['nodes'], list) or []
            if entry.get('__typename') == 'Video']
```
Also, not sure if the isinstance check is actually needed

File: yt_dlp/extractor/ted.py [hunk 5]
```diff
@@ -149,84 +90,76 @@ def _real_extract(self, url):
...
-        else:
+        elif m.group('type_playlist'):
             return self._playlist_videos_info(url, name)
+        else:
+            return self._series_videos_info(url, name, m.group('season'))
 
     def _playlist_videos_info(self, url, name):
         '''Returns the videos of the playlist'''
+        webpage = self._download_webpage(url, name, 'Downloading playlist webpage')
+        json = self._extract_info(webpage, 'json')
```
Comment (pukkandan) [inline yt_dlp/extractor/ted.py:101]: use a diff var name

File: yt_dlp/extractor/ted.py [hunk 6]
```diff
@@ -149,84 +90,76 @@ def _real_extract(self, url):
...
-                if finfo:
-                    f.update(finfo)
-
-        player_talk = talk_info['player_talks'][0]
+        playlist_entries = []
+        if season:
+            for s in [playlist for playlist in seasonlist if playlist['seasonNumber'] == season]:
+                playlist_entries.extend(self._parse_playlist(s))
+        else:
+            playlist_entries = self._parse_playlist(seasonlist[0])
```
Comment (pukkandan) [inline yt_dlp/extractor/ted.py:126]: shouldn't all seasons be downloaded if no number was given?

File: yt_dlp/extractor/ted.py [hunk 7]
```diff
@@ -149,84 +90,76 @@ def _real_extract(self, url):
...
-            for f in formats:
-                finfo = self._NATIVE_FORMATS.get(f['format_id'].split('-')[0])
-                if finfo:
-                    f.update(finfo)
-
-        player_talk = talk_info['player_talks'][0]
+        playlist_entries = []
+        if season:
+            for s in [playlist for playlist in seasonlist if playlist['seasonNumber'] == season]:
+                playlist_entries.extend(self._parse_playlist(s))
```
Comment (pukkandan) [inline yt_dlp/extractor/ted.py:124]:
```suggestion
            playlist_entries = [self._parse_playlist(s) for s in seasonlist if s.get('seasonNumber') == season]
```

File: yt_dlp/extractor/ted.py [hunk 8]
```diff
@@ -149,84 +90,76 @@ def _real_extract(self, url):
...
+        talk_info = try_get(json, lambda x: x['props']['pageProps']['videoData'], dict)
+
+        video_id = talk_info.get('id')
+        title = talk_info.get('title') or self._og_search_title(webpage)
+        uploader = talk_info.get('presenterDisplayName')
+        release_date = unified_strdate(talk_info.get('recordedOn'))
+        upload_date = unified_strdate(talk_info.get('publishedAt'))
+        view_count = str_to_int(talk_info.get('viewedCount'))
+        duration = parse_duration(talk_info.get('duration')) or parse_duration(self._og_search_property('video:duration', webpage))
+        description = parse_duration(talk_info.get('description')) or self._og_search_description(webpage)
```
Comment (pukkandan) [inline yt_dlp/extractor/ted.py:145]: do u really need all these vars, or can u inline a lot of then in the return (or atleast move them close to return)

File: yt_dlp/extractor/ted.py [hunk 9]
```diff
@@ -149,84 +90,76 @@ def _real_extract(self, url):
...
                     continue
                 stream_url = url_or_none(resources.get('stream'))
                 if not stream_url:
                     continue
-                formats.extend(self._extract_m3u8_formats(
+                m3u8, subtitles = self._extract_m3u8_formats_and_subtitles(
                     stream_url, video_name, 'mp4', m3u8_id=format_id,
-                    fatal=False))
+                    fatal=False)
+                formats.extend(m3u8)
```
Comment (pukkandan) [inline yt_dlp/extractor/ted.py:162]:
```suggestion
                m3u8_formats, m3u8_subs = self._extract_m3u8_formats_and_subtitles(
                    stream_url, video_name, 'mp4', m3u8_id=format_id,
                    fatal=False)
                formats.extend(m3u8_formats)
                subtitles = self._merge_subtitles(subtitles, m3u8_subs)
```

File: yt_dlp/extractor/ted.py [hunk 10]
```diff
@@ -300,68 +233,25 @@ def _talk_info(self, url, video_name):
...
-        description = self._html_search_regex(
-            [
-                r'(?s)<h4 class="[^"]+" id="h3--about-this-talk">.*?</h4>(.*?)</div>',
-                r'(?s)<p><strong>About this talk:</strong>\s+(.*?)</p>',
-            ],
-            webpage, 'description', fatal=False)
+        # tags
+        tags = try_get(playerData, lambda x: x['targeting']['tag'], str)
+        if tags:
+            tags = tags.split(',')
```
Comment (pukkandan) [inline yt_dlp/extractor/ted.py:242]:
```suggestion
        tags = try_get(playerData, lambda x: x['targeting']['tag'].split(','))
```
can be inlined in return

File: yt_dlp/extractor/ted.py [hunk 11]
```diff
@@ -300,68 +233,25 @@ def _talk_info(self, url, video_name):
 
         self._sort_formats(formats)
 
-        video_id = compat_str(talk_info['id'])
+        # trim thumbnail resize parameters
+        thumbnail = self._search_regex(r'^(http[^?]*)', playerData.get('thumb'), 'thumbnail', default=None) or self._search_regex(r'^(http[^?]*)', self._og_search_property('image', webpage), 'thumbnail', default=None)
```
Comment (pukkandan) [inline yt_dlp/extractor/ted.py:237]:
```suggestion
        thumbnail = playerData.get('thumb') or self._og_search_property('image', webpage)
        if thumbnail:
            # trim thumbnail resize parameters
            thumbnail = thumbnail.split('?')[0]
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/ted.py (hunk 7 above)
Comment (trassshhub) [reply yt_dlp/extractor/ted.py]: I am getting errors with your suggested change.
Still haven't got my head around the Python for loop short forms, please check if this change below is acceptable.

https://github.com/trassshhub/yt-dlp/blob/6ea4d15a3d79dd9f504a8ca242c8c40a0cf2c080/yt_dlp/extractor/ted.py#L114-L118

Comment (pukkandan) [reply yt_dlp/extractor/ted.py]: let it be as-is for now

**Comment by pukkandan** [comment]
Add some tests for playlist, series and #season

**Comment by trassshhub** [comment]
I was trying to use series_id instead of id, it works with the binary but the test is complaining that the id key is not present.
Is there a way around this or shall I just leave it to use id?

File: yt_dlp/extractor/ted.py [hunk 12]
```diff
@@ -26,120 +21,114 @@ class TEDIE(InfoExtractor):
...
             |
-            ((?P<type_talk>talks)) # We have a simple talk
+            (?P<type_series>series) # We have a series
             |
-            (?P<type_watch>watch)/[^/]+/[^/]+
+            (?P<type_talk>talks) # We have a simple talk
         )
         (/lang/(.*?))? # The url may contain the language
-        /(?P<name>[\w-]+) # Here goes the name and then ".html"
+        /(?P<name>[\w-]+)(?:\#season_(?P<season>\d))? # Here goes the name
```
Comment (Ashish0804) [inline yt_dlp/extractor/ted.py:29]: Can you remove these comments?
They are completely unnecessary and annoying....idk how they were allowed to be merged last time

File: yt_dlp/extractor/ted.py [hunk 13]
```diff
@@ -26,120 +21,114 @@ class TEDIE(InfoExtractor):
...
 
-    def _extract_info(self, webpage):
-        info_json = self._search_regex(
-            r'(?s)q\(\s*"\w+.init"\s*,\s*({.+?})\)\s*</script>',
-            webpage, 'info json')
-        return json.loads(info_json)
+    def _parse_playlist(self, playlist):
+        playlist_entries = []
+
+        for entry in try_get(playlist, lambda x: x['videos']['nodes'], list):
```
Comment (Ashish0804) [inline yt_dlp/extractor/ted.py:127]:
```suggestion
        for entry in try_get(playlist, lambda x: x['videos']['nodes'], list) or []:
```

File: yt_dlp/extractor/ted.py [hunk 14]
```diff
@@ -26,120 +21,114 @@ class TEDIE(InfoExtractor):
...
+            if entry.get('__typename') == 'Video':
+                playlist_entries.append(self.url_result(entry.get('canonicalUrl'), self.ie_key()))
```
Comment (Ashish0804) [inline yt_dlp/extractor/ted.py:129]:
```suggestion
            if entry.get('__typename') == 'Video' and entry.get('canonicalUrl'):
                playlist_entries.append(self.url_result(entry['canonicalUrl'], self.ie_key()))
```

File: yt_dlp/extractor/ted.py [hunk 15]
```diff
@@ -149,84 +138,70 @@ def _real_extract(self, url):
         name = m.group('name')
         if m.group('type_talk'):
             return self._talk_info(url, name)
-        elif m.group('type_watch'):
-            return self._watch_info(url, name)
-        else:
+        elif m.group('type_playlist'):
             return self._playlist_videos_info(url, name)
+        else:
+            return self._series_videos_info(url, name, m.group('season'))
```
Comment (Ashish0804) [inline yt_dlp/extractor/ted.py:144]: use `elif` if possible

File: yt_dlp/extractor/ted.py [hunk 16]
```diff
@@ -149,84 +138,70 @@ def _real_extract(self, url):
...
 
     def _playlist_videos_info(self, url, name):
         '''Returns the videos of the playlist'''
```
Comment (Ashish0804) [inline yt_dlp/extractor/ted.py:147]:
```suggestion
```

File: yt_dlp/extractor/ted.py [hunk 17]
```diff
@@ -149,84 +138,70 @@ def _real_extract(self, url):
...
-        for entry in re.findall(r'(?s)<[^>]+data-ga-context=["\']playlist["\'][^>]*>', webpage):
-            attrs = extract_attributes(entry)
-            entry_url = compat_urlparse.urljoin(url, attrs['href'])
-            playlist_entries.append(self.url_result(entry_url, self.ie_key()))
-
-        final_url = self._og_search_url(webpage, fatal=False)
-        playlist_id = (
-            re.match(self._VALID_URL, final_url).group('playlist_id')
-            if final_url else None)
+        playlist = try_get(info, lambda x: x['props']['pageProps']['playlist'], dict)
```
Comment (Ashish0804) [inline yt_dlp/extractor/ted.py:151]:
```suggestion
        playlist = try_get(info, lambda x: x['props']['pageProps']['playlist'], dict) or {}
```

File: yt_dlp/extractor/ted.py [hunk 18]
```diff
@@ -149,84 +138,70 @@ def _real_extract(self, url):
...
+            playlist_title=playlist.get('title') or self._og_search_title(webpage, fatal=False).replace(' | TED Talks', ''),
             playlist_description=self._og_search_description(webpage))
 
-    def _talk_info(self, url, video_name):
-        webpage = self._download_webpage(url, video_name)
+    def _series_videos_info(self, url, name, season):
+        '''Returns the videos of the series playlist'''
+        webpage = self._download_webpage(url, name, 'Downloading series webpage')
+        info = self._extract_info(webpage, 'json')
+        series = try_get(info, lambda x: x['props']['pageProps']['series'])
```
Comment (Ashish0804) [inline yt_dlp/extractor/ted.py:164]:
```suggestion
        series = try_get(info, lambda x: x['props']['pageProps']['series']) or {}
```

**Comment by Ashish0804** [comment]
This was a partial review.
Remove useless/obvious comments and add `or <empty object>` after every try_get and .get used to get iterables.

**Comment by trassshhub** [comment]
Thanks for the review, I have committed the suggested changes.

by the way, for this issue I mentioned above, it may not be directly related to this PR but I want some insights from the reviewers.

> I was trying to use series_id instead of id, it works with the binary but the test is complaining that the id key is not present. Is there a way around this or shall I just leave it to use id?
> 
> ```
> Traceback (most recent call last):
>   File "P:\BT\software\yt-dlp\test\test_download.py", line 191, in test_template
>     expect_info_dict(self, res_dict, test_case.get('info_dict', {}))
>   File "P:\BT\software\yt-dlp\test\helper.py", line 257, in expect_info_dict
>     test_info_dict = sanitize_got_info_dict(got_dict)
>   File "P:\BT\software\yt-dlp\test\helper.py", line 238, in sanitize_got_info_dict
>     if test_info_dict.get('display_id') == test_info_dict['id']:
> KeyError: 'id'
> ```

common.py states that playlist 'can' have 'id' which means it is not compulsory.
https://github.com/yt-dlp/yt-dlp/blob/32b95bb6436fd17ee21cd55217f16361d4e7494a/yt_dlp/extractor/common.py#L385-L395

But the test helper seems to be treating 'id' as compulsory and therefore using test_info_dict['id'] and not test_info_dict.get('id')?
https://github.com/yt-dlp/yt-dlp/blob/32b95bb6436fd17ee21cd55217f16361d4e7494a/test/helper.py#L237-L239

Which should be the correct behavior?

**Comment by pukkandan** [comment]
Probably an oversight. I'll look into it. Playlist Id is optional

**Pull request merged**